### PR TITLE
Ubuntu Advantage config: http_proxy, https_proxy

### PR DESCRIPTION
**Pull request by chifac08**
## Proposed Commit Message
Ubuntu Advantage config: http_proxy, https_proxy

```
## Summary
Implemented the usage of

- http_proxy
- https_proxy

parameters for Ubuntu Advantage. Parameter will be set before an attachment takes place.

```

## Test Steps
I did my best in writing unit tests and would be delighted if you add more as I am not very familiar with that.
Following scenarios are tested:

- UA config with weird params
- UA config with an invalid url
- UA config with non string values for keys

## Checklist:
 - [ x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [ x] I have updated or added any unit tests accordingly
 - [ x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/config/schemas/schema-cloud-config-v1.json [hunk 1]
```diff
@@ -1974,6 +1974,20 @@
...
               "type": "string",
               "description": "Required contract token obtained from https://ubuntu.com/advantage to attach."
+            },
+            "config": {
+              "type": "object",
+              "description": "Configuration settings or override Ubuntu Advantage config",
+              "properties": {
+                "http_proxy": {
+                  "type": "string",
+                  "description": "Ubuntu Advantage HTTP Proxy"
```
Comment (blackboxsw) [inline cloudinit/config/schemas/schema-cloud-config-v1.json:1984]:
```suggestion
                  "format": "uri",
                  "description": "Ubuntu Advantage HTTP Proxy URL"
```

File: cloudinit/config/schemas/schema-cloud-config-v1.json [hunk 2]
```diff
@@ -1974,6 +1974,20 @@
...
+                },
+                "https_proxy": {
+                  "type": "string",
+                  "description": "Ubuntu Advantage HTTPS Proxy"
```
Comment (blackboxsw) [inline cloudinit/config/schemas/schema-cloud-config-v1.json:1988]:
```suggestion
                  "format": "uri",
                  "description": "Ubuntu Advantage HTTPS Proxy URL"
```

File: cloudinit/config/schemas/schema-cloud-config-v1.json [hunk 3]
```diff
@@ -1974,6 +1974,20 @@
...
+                }
```
Comment (blackboxsw) [inline cloudinit/config/schemas/schema-cloud-config-v1.json:1989]: I'd like us to add all known proxy config options to this schema here as they seem the most likely customers would want to adapt and unlikely to change name in the near future: http_proxy, https_proxy,  apt_http_proxy, apt_https_proxy,  ua_apt_http_proxy ,  ua_apt_https_proxy , global_apt_http_proxy, global_apt_https_proxy

```suggestion
                },
                "global_apt_http_proxy": {
                  "type": "string",
                  "format": "uri",
                  "description": "HTTP Proxy URL used for all APT repositories on a system. Stored at ``/etc/apt/apt.conf.d/90ubuntu-advantage-aptproxy``"
               },
                "global_apt_https_proxy": {
                  "type": "string",
                  "format": "uri",
                  "description": "HTTPS Proxy URL used for all APT repositories on a system. Stored at ``/etc/apt/apt.conf.d/90ubuntu-advantage-aptproxy``"
               },
                "ua_apt_http_proxy": {
                  "type": "string",
                  "format": "uri",
                   "description": "HTTP Proxy URL used only for Ubuntu Advantage APT repositories. Stored at ``/etc/apt/apt.conf.d/90ubuntu-advantage-aptproxy``"
               },
                "ua_apt_https_proxy": {
                  "type": "string",
                  "format": "uri",
                  "description": "HTTP Proxy URL used only for Ubuntu Advantage APT repositories. Stored at ``/etc/apt/apt.conf.d/90ubuntu-advantage-aptproxy``"               },
```

File: cloudinit/config/cc_ubuntu_advantage.py [hunk 4]
```diff
@@ -69,6 +69,19 @@
...
         ),
+        dedent(
+            """\
+        # Set a http(s) proxy before attaching the machine to an
+        # Ubuntu Advantage support contract and enabling the FIPS service.
+        ubuntu_advantage:
+          token: <ua_contract_token>
+          config:
+            http_proxy: <http_proxy>
+            https_proxy: <https_proxy>
```
Comment (blackboxsw) [inline cloudinit/config/cc_ubuntu_advantage.py:80]: Probably want to provide a fake url format to give a more concrete example. Eventually our unit tests will valid URI format.
Also let's add the other proxy examples
```suggestion
            https_proxy: 'https://some-proxy:8088'
            global_apt_https_proxy: 'http://some-global-apt-proxy:8088/'
            global_apt_http_proxy: 'https://some-global-a'pt-proxy:8088/'          
            ua_apt_http_proxy: 'http://10.0.10.10:3128'
            ua_apt_https_proxy: 'https://10.0.10.10:3128'
```

File: tools/.github-cla-signers [hunk 5]
```diff
@@ -18,6 +18,7 @@ bmhughes
 candlerb
 cawamata
 cclauss
+chifac08
```
Comment (blackboxsw) [inline tools/.github-cla-signers:21]: I confirmed this user has signed the CLA. Welcome to cloud-init @chifac08 and thanks!

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Thank you much @chifac08 for supplying this PR and making cloud-init better. 
Sorry for the delay, it's been a busy week here. I've given an initial review with comments, I like the additional verification you are doing. I need to perform a couple of tests Monday and I can finalize my review.  I think it's worth us hammering out a bit more of the *apt*proxy variables (and minor schema docs)

Thanks again for the great contribution here.

Note the global_* and ua_* proxy config variables are nearly queued for a public release version 27.9 for ubuntu-advantage-tools. I expect this SRU from that team will land before cloud-init releases this PR upstream.

One can see `add-apt-repostory ppa:ua-client/daily; apt install ubuntu-advantage-tools` to see those ua config settings.

**Comment by chifac08** [comment]
Hi @blackboxsw,

thank you for reviewing my PR and sorry the late reply. 

I implemented and tested the changes you requested. 

thx

File: cloudinit/config/cc_ubuntu_advantage.py [hunk 6]
```diff
@@ -102,6 +150,44 @@ def configure_ua(token=None, enable=None):
...
+    enable_errors = []
+
+    # UA Config
+    for key, value in sorted(config.items()):
+        if value is None:
+            LOG.debug("Unsetting UA config for %s", key)
+            config_cmd = ["ua", "config", "unset", key]
+        else:
+            LOG.debug("Setting UA config %s=%s", key, value)
+            config_cmd = ["ua", "config", "set", f"{key}='{value}'"]
```
Comment (blackboxsw) [inline cloudinit/config/cc_ubuntu_advantage.py:172]: I know I suggested this, but in testing your PR, it looks like we shouldn't provide the single-quotes for these proxy URLs  as ua set config *proxy='http://someurl' will fail to parse the rendered command and reject the full value 'http://someurl' as an invalid URL.

We should probably do the following only if we have whitespace in the value:
```suggestion
            if re.search(r"\s", value):
                key_value = f"{key}={re.escape(value)}"
            else:
                key_value = f"{key}={value}"
            config_cmd = ["ua", "config", "set", key_value]
```

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Thanks for the updates. A  couple of minor inline comments, some were missed in the initial review.
 *  I think it's worth use using urlparse too to validate the string value. We can then provide a concise error from cloud-init
 *  We need to avoid single quoting the URL value when calling ua config set *proxy='http://yoururl' as that is being misinterpreted on the cmdline. Maybe only escape the string if the value contains whitespace.

Here's the consolidated diff of my comments and an extension to the unit test to expose the non-URL string case for your review.
```diff

diff --git a/cloudinit/config/cc_ubuntu_advantage.py b/cloudinit/config/cc_ubuntu_advantage.py
index 2aa8db037..4fcd9055f 100644
--- a/cloudinit/config/cc_ubuntu_advantage.py
+++ b/cloudinit/config/cc_ubuntu_advantage.py
@@ -2,7 +2,9 @@
 
 """ubuntu_advantage: Configure Ubuntu Advantage support services"""
 
+import re
 from textwrap import dedent
+from urllib.parse import urlparse
 
 from cloudinit import log as logging
 from cloudinit import subp, util
@@ -108,18 +110,26 @@ def supplemental_schema_validation(ua_config):
     errors = []
     nl = "\n"
     for key, value in sorted(ua_config.items()):
-        if (
-            key == "http_proxy"
-            or key == "https_proxy"
-            or key == "global_apt_http_proxy"
-            or key == "global_apt_https_proxy"
-            or key == "ua_apt_http_proxy"
-            or key == "ua_apt_https_proxy"
+        if key in (
+            "http_proxy",
+            "https_proxy",
+            "global_apt_http_proxy",
+            "global_apt_https_proxy",
+            "ua_apt_http_proxy",
+            "ua_apt_https_proxy",
         ):
-            if not isinstance(value, str):
+            try:
+                parsed_url = urlparse(value)
+                if parsed_url.scheme not in ("http", "https"):
+                    errors.append(
+                        f"Expected URL scheme http/https for ua:config:{key}."
+                        f" Found: {value}"
+                    )
+            except (AttributeError, ValueError):
                 errors.append(
-                    f"Expected a url for ua:config:{key}. Found: {value}"
+                    f"Expected a URL for ua:config:{key}. Found: {value}"
                 )
+
     if errors:
         raise ValueError(
             f"Invalid ubuntu_advantage configuration:{nl}{nl.join(errors)}"
@@ -169,7 +179,11 @@ def configure_ua(token=None, enable=None, config=None):
             config_cmd = ["ua", "config", "unset", key]
         else:
             LOG.debug("Setting UA config %s=%s", key, value)
-            config_cmd = ["ua", "config", "set", f"{key}='{value}'"]
+            if re.search(r"\s", value):
+                key_value = f"{key}={re.escape(value)}"
+            else:
+                key_value = f"{key}={value}"
+            config_cmd = ["ua", "config", "set", key_value]
 
         try:
             subp.subp(config_cmd)
@@ -178,16 +192,16 @@ def configure_ua(token=None, enable=None, config=None):
 
     if enable_errors:
         for param, error in enable_errors:
-            msg = 'Failure enabling "{param}":\n{error}'.format(
-                param=param, error=str(error)
+            LOG.warning(
+                'Failure enabling "{param}":\n{error}'.format(
+                    param=param, error=str(error)
+                )
             )
-            util.logexc(LOG, msg)
         raise RuntimeError(
             "Failure enabling Ubuntu Advantage config(s): {}".format(
                 ", ".join('"{}"'.format(param) for param, _ in enable_errors)
             )
         )
-    enable_errors = []
     attach_cmd = ["ua", "attach", token]
     LOG.debug("Attaching to Ubuntu Advantage. %s", " ".join(attach_cmd))
     try:
diff --git a/tests/unittests/config/test_cc_ubuntu_advantage.py b/tests/unittests/config/test_cc_ubuntu_advantage.py
index 7e9a399cf..cbac957b8 100644
--- a/tests/unittests/config/test_cc_ubuntu_advantage.py
+++ b/tests/unittests/config/test_cc_ubuntu_advantage.py
@@ -206,10 +206,16 @@ class TestConfigureUA(CiTestCase):
 
     def test_configure_ua_config_error_non_string_values(self):
         """ValueError raised for any values expected as string type."""
-        cfg = {"http_proxy": ["no-proxy"], "https_proxy": ["no-proxy2"]}
+        cfg = {
+            "global_apt_http_proxy": "noscheme",
+            "http_proxy": ["no-proxy"],
+            "https_proxy": 1,
+        }
         errors = [
-            "Expected a url for ua:config:http_proxy. Found: ['no-proxy']",
-            "Expected a url for ua:config:https_proxy. Found: ['no-proxy2']",
+            "Expected URL scheme http/https for"
+            " ua:config:global_apt_http_proxy. Found: noscheme",
+            "Expected a URL for ua:config:http_proxy. Found: ['no-proxy']",
+            "Expected a URL for ua:config:https_proxy. Found: 1",
         ]
         with self.assertRaises(ValueError) as context_manager:
             supplemental_schema_validation(cfg)
```
I approve once we minimally revmo

**Comment by blackboxsw** [comment]
@chifac08 I'm good to land this PR once we minimally avoid single quoting the config values passed to the `ua config set key=value` commandline.

**Comment by chifac08** [comment]
Hi @blackboxsw 

thanks you for testing and reviewing my PR. I definitely like the above mentioned changes, especially the one to validate a url with urlparse rather than validating if it's a string or not. Please feel free to push those changes to this PR and hopefully we are ready to go :)

**Comment by blackboxsw** [comment]
@chifac08 just pushed the review comment diff. We'll merge once CI succeeds.

**Pull request merged**